src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	for (int i = 0; i < p_val->c_type_parameters; i++)
+	{
+		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "type: %d\n", ((MonoType*)p_val->rg_type_parameters[i])->type);
+	}
+
+	// The type fits into the batch, so update our state
+	p_type_logger->m_n_bulk_type_value_count++;
+	p_type_logger->m_n_bulk_type_value_byte_count += cb_val;
+	return p_type_logger->m_n_bulk_type_value_count - 1;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// High-level method to batch a type and (recursively) its type parameters, flushing to
+// ETW as needed.  This is called by ep_rt_mono_log_type_and_parameters_if_necessary.
+//
+// Arguments:
+//      * mono_type - MonoType to be logged
+//      * type_id - Unique type identifier of mono_type to batch
+
+void
+ep_rt_mono_log_type_and_parameters (BulkTypeEventLogger *p_type_logger, MonoType *mono_type, intptr_t type_id)
+{
+	// Batch up this type.  This grabs useful info about the type, including any
+	// type parameters it may have, and sticks it in m_rg_bulk_type_values
+	uint32_t i_bulk_type_event_data = ep_rt_mono_log_single_type (p_type_logger, mono_type, type_id);
+	if (i_bulk_type_event_data == -1)
+	{
+		// There was a failure trying to log the type, so don't bother with its type
+		// parameters
+		return;
+	}
+
+	// Look at the type info we just batched, so we can get the type parameters
+	BulkTypeValue *p_val = &p_type_logger->m_rg_bulk_type_values[i_bulk_type_event_data];
+
+	// We're about to recursively call ourselves for the type parameters, so make a
+	// local copy of their type handles first (else, as we log them we could flush
+	// and clear out m_rg_bulk_type_values, thus trashing p_val)
+	uint32_t c_params = p_val->c_type_parameters;
+	intptr_t rg_type_parameters[c_params];
+	MonoType *rg_mono_type_parameters[c_params];
+	for (uint32_t i = 0; i < c_params; i++)
+	{
+		rg_type_parameters[i] = p_val->rg_type_parameters[i];
+		rg_mono_type_parameters[i] = p_val->rg_mono_type_parameters[i];
+	}
+
+	for (uint32_t i = 0; i < c_params; i++)
+		ep_rt_mono_log_type_and_parameters_if_necessary (p_type_logger, rg_mono_type_parameters[i], rg_type_parameters[i]);
+}
+
+//---------------------------------------------------------------------------------------
+//
+// Outermost level of ETW-type-logging.  This method is used to log a unique type identifier
+// (in this case a MonoType) and (recursively) its type parameters when present.
+//
+// Arguments:
+//      * mono_type - MonoType to be logged
+//      * type_id - Unique type identifier of mono_type
+
+// static
+void
+ep_rt_mono_log_type_and_parameters_if_necessary (BulkTypeEventLogger *p_type_logger, MonoType *mono_type, intptr_t type_id)
+{
+	// TODO Log the type if necessary
```

Comment:
It looks like SendMethodDetailsEvent in CoreCLR always logs the type with `ETW::TypeSystemLog::kTypeLogBehaviorAlwaysLog` for the two times it calls `ETW::TypeSystemLog::LogTypeAndParametersIfNecessary` https://github.com/dotnet/runtime/blob/5992145db2cb57956ee444aa0f0c2f3f85ee3673/src/coreclr/vm/eventtrace.cpp#L6442 and https://github.com/dotnet/runtime/blob/5992145db2cb57956ee444aa0f0c2f3f85ee3673/src/coreclr/vm/eventtrace.cpp#L6468, being consumed at https://github.com/dotnet/runtime/blob/5992145db2cb57956ee444aa0f0c2f3f85ee3673/src/coreclr/vm/eventtrace.cpp#L3275-L3280 and the `TypeLogBehavior` is recursively passed through. Since this PR aims to primarily implement `ep_rt_mono_send_method_details_event` and `ep_rt_mono_fire_bulk_type_event`, where the latter is currently only called by the former, the hash map wouldn't effectively be used unless something else called `ep_rt_mono_log_type_and_parameters_if_necessary`, which I'm not certain we have plans for that yet. Would it be better to just implement the `if necessary` when its actually used so it can be tested properly?